Bump minimum Go version to 1.24.0

axw opened the pull request:
#### Description

[Go 1.25.0](https://go.dev/doc/devel/release#go1.25.0) is released, time to bump the minimum supported version.

See also https://github.com/open-telemetry/opentelemetry-collector/pull/13627

#### Link to tracking issue

Fixes #38970 

#### Testing

N/A

#### Documentation

Updated

Commit:
Bump minimum Go version to 1.24.0

axw commented:
Requires https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41966

Commit:
Add changelog

axw commented:
Going back to draft, there's a whole lot of lint errors. If it's going to take a long time to fix, it might be worth temporarily disabling them and fixing in a followup, to avoid blocking the version bump.

Commit:
Fix golangci-lint errors

- Use testing.T.Context
- Use testing.T.Chdir

florianl commented on cmd/codecovgen/go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/open-telemetry/opentelemetry-collector-contrib/cmd/codecovgen
 
-go 1.23.0
+go 1.24
```
Is there a specific reason to drop the patch version?
My understanding is, that `go 1.24` will always point to the latest patch version. While `go 1.23.0` always pointed to the `0`-patch version.
As security issues are often addressed in patch versions, this would make a difference. On the other hand, if issues related to Go are encountered, it will be easier if a patch version is chosen explicitly.

mx-psi replied:
> My understanding is, that go 1.24 will always point to the latest patch version. While go 1.23.0 always pointed to the 0-patch version.

I don't think this is the case, per https://go.dev/doc/toolchain#version

> Released versions of Go use the version syntax ‘1.N.P’, denoting the Pth release of Go 1.N. The initial release is 1.N.0, like in ‘1.21.0’. Later releases like 1.N.9 are often referred to as patch releases.
>
> Go 1.N release candidates, which are issued before 1.N.0, use the version syntax ‘1.NrcR’. The first release candidate for Go 1.N has version 1.Nrc1, like in 1.23rc1.
>
> The syntax ‘1.N’ is called a “language version”. It denotes the overall family of Go releases implementing that version of the Go language and standard library.
> 
> [...] Within a language version, the ordering from least to greatest is: the language version itself, then release candidates ordered by R, then releases ordered by P.
>
> For example, **1.21 < 1.21rc1 < 1.21rc2 < 1.21.0 < 1.21.1 < 1.21.2.**

So the difference between `go 1.24` and `go 1.24.0` is that `go 1.24` claims to support any version within 1.24.x, including RCs, while `go 1.24.0` only supports stable versions of the 1.24.x release series.

Official Go packages are using the `go 1.24.0` syntax, see e.g. https://cs.opensource.google/go/x/crypto/+/master:go.mod, which makes it hard in practice to use the `go 1.24` syntax (we could use it for now but it won't work for long IMO).

> As security issues are often addressed in patch versions, this would make a difference

I don't think this has any security impact: the `go` directive just sets the minimum version supported, but you can use any Go compiler version to compile a binary using that module that is equal or higher to the `go` directive.

> The real reason I changed it though is because it seems that having the .0 on the end affects the toolchain version selection -- without this change, govulncheck was failing due to some vulnerabilities that existed in go 1.24.0. I don't totally understand this, so it's possible that I've done something wrong.

An alternative to this would be to bump the job handling otelcorecol to 1.25, since starting on 1.25:

> When the go command updates the go line in a go.mod or go.work file, it [no longer](https://tip.golang.org/ref/mod#go-mod-file-toolchain) adds a toolchain line specifying the command’s current version.

florianl replied:
I'm happy to see the bump to Go 1.24. Just wanted to raise awareness of this minor detail. And sorry for mixing the purpose of the patch version :pray:

Commit:
Revert some uses of testing.T.Context

The context is cancelled before cleanups,
so we can't use it inside cleanup functions.

Commit:
make generate tidy

Commit:
datadogexporter: work around data race

The exporter appears to have a data race between
traces being sent and the exporter shutting down,
related to lazy initialisation of the obfuscator.

atoulme committed:
Merge branch 'main' into go1.24

Commit:
More usetesting lint fixes

axw commented:
The govulncheck errors are also on main, I don't think we should try to fix them in this PR - it's already a bit extreme as it is.
